File: 추천50제/완전탐색.백트래킹/14888.py
import sys
input = sys.stdin.readline

n = int(input())
array = list(map(int,input().split()))
operator = list(map(int,input().split()))

answer = []
def dfs(i,number):
    if i==n:
        answer.append(number)
        return
    if operator[0]>0:
        operator[0]-=1
        dfs(i+1,number+array[i])
        operator[0]+=1

    if operator[1]>0:
        operator[1]-=1
        dfs(i+1,number-array[i])
        operator[1]+=1

    if operator[2]>0:
        operator[2]-=1
        dfs(i+1,number*array[i])
        operator[2]+=1

    if operator[3]>0:
        operator[3]-=1
        dfs(i+1,int(number/array[i]))
        operator[3]+=1

dfs(1,array[0])
print(max(answer))
print(min(answer))

File: BOJ/14888.js
let fs = require('fs');
let input = fs.readFileSync('./test.txt').toString().trim().split('\n');
let n = Number(input.shift());
let arr = input.shift().split(' ').map(Number);
let operator = input.shift().split(' ').map(Number);
let Max_answer = -Infinity;
let Min_answer = Infinity;
const recursion = (cur_value, depth) => {
  if (depth == n) {
    Max_answer = Math.max(Max_answer, cur_value);
    Min_answer = Math.min(Min_answer, cur_value);
    return;
  }
  for (let i = 0; i < 4; i++) {
    if (operator[i]) {
      operator[i] -= 1;
      switch (i) {
        case 0:
          recursion(cur_value + arr[depth], depth + 1);
          break;
        case 1:
          recursion(cur_value - arr[depth], depth + 1);
          break;
        case 2:
          recursion(cur_value * arr[depth], depth + 1);
          break;
        case 3:
          recursion(Math.trunc(cur_value / arr[depth]), depth + 1);
          break;
      }
      operator[i] += 1;
    }
  }
};
recursion(arr[0], 1);
console.log(parseInt(Max_answer));
console.log(parseInt(Min_answer));
//-0 을 주의하자.
